ymer/src/main/java/com/avanza/ymer/PersistedInstanceIdRecalculationService.java

```diff
@@ -57,34 +97,92 @@ public void recalculatePersistedInstanceId(String collectionName, int numberOfIn
 		if (!indexDropped) {
 			log.info("Step 1/3\tNo index to drop");
 		}
-		AtomicInteger count = new AtomicInteger();
-		collection.findByQuery(query(where(DOCUMENT_ROUTING_KEY).exists(true)))
-				.peek(it -> {
-					int currentCount = count.incrementAndGet();
-					if (currentCount % 10_000 == 0) {
-						log.info("Step 2/3\tUpdated persisted instance id for {} documents", currentCount);
+
+		log.info("Using batch size {}", BATCH_SIZE);
+		try (Stream<List<Document>> batches = StreamUtils.buffer(collection.findByQuery(createQuery(BATCH_SIZE)), BATCH_SIZE)) {
+			LongAdder analyzedCount = new LongAdder();
+			AtomicInteger updatedCount = new AtomicInteger();
+			batches.forEach(batch -> collection.bulkWrite(bulkWriter -> {
+				Map<Integer, List<Document>> updatesByInstanceId = batch.stream()
+						.collect(groupingBy(it -> getInstanceId(it.get(DOCUMENT_ROUTING_KEY), numberOfPartitions)));
+
+				updatesByInstanceId.forEach((instanceId, documents) -> {
+					Set<Object> ids = documents.stream()
+							.peek(it -> analyzedCount.increment())
+							.filter(document -> !Objects.equals(instanceId, document.get(DOCUMENT_INSTANCE_ID)))
+							.map(document -> document.get("_id"))
+							.filter(Objects::nonNull)
+							.peek(it -> {
+								int currentCount = updatedCount.incrementAndGet();
+								if (currentCount % 10_000 == 0) {
+									log.info("Step 2/3\tUpdated persisted instance id for {} documents ({} analyzed)", currentCount, analyzedCount.sum());
+								}
+							})
+							.collect(toSet());
+					if (!ids.isEmpty()) {
+						bulkWriter.updatePartialByIds(ids, Map.of(DOCUMENT_INSTANCE_ID, instanceId));
 					}
-				})
-				.forEach(document -> {
-					Object id = document.get("_id");
-					int instanceId = getInstanceId(document.get(DOCUMENT_ROUTING_KEY), numberOfInstances);
-					collection.updateById(id, Map.of(DOCUMENT_INSTANCE_ID, instanceId));
 				});
-
-		log.info("Step 2/3\tUpdated persisted instance id for {} documents", count.get());
+			}));
+			log.info("Step 2/3\tUpdated persisted instance id for {} documents total ({} analyzed total)", updatedCount.get(), analyzedCount.sum());
+		}
 
 		boolean indexExists = collection.getIndexes()
 				.filter(index -> index.isIndexForFields(Set.of(DOCUMENT_INSTANCE_ID)))
 				.anyMatch(index -> index.getName().endsWith(indexSuffix));
 		if (indexExists) {
 			log.info("Step 3/3\tIndex does not need to be created because it already exists");
 		} else {
-			String indexName = "_index" + DOCUMENT_INSTANCE_ID + "_numberOfInstances" + indexSuffix;
+			String indexName = "_index" + DOCUMENT_INSTANCE_ID + "_numberOfPartitions" + indexSuffix;
 			log.info("Step 3/3\tCreating index {}", indexName);
 			IndexOptions options = new IndexOptions().name(indexName).background(true);
```

Comment:
Does `.background(true)` mean that the mongodb server will create the index in the background after this statement has completed? If so, maybe the logging below should be changed from `Done creating index` to something that indicates that it will be done some time in the future? Or is it possible for this method to know when the index is completely created and only log completion once this is done?

Reply:
Creating the index will block until it is done. Background creation means that it will not block other operations to the collection.
See https://docs.mongodb.com/v3.6/core/index-creation/#index-creation-background
> Background indexing operations run in the background so that other database operations can run while creating the index. However, the mongo shell session or connection where you are creating the index will block until the index build is complete.